packages/web5/lib/src/dids/did_method_resolver.dart

```diff
@@ -1,13 +1,37 @@
-import 'package:web5/src/dids/did_resolution_result.dart';
+import 'package:web5/src/dids/data_models.dart';
 
 /// Represents a method resolver for a specific DID method.
 class DidMethodResolver {
   /// The name of the DID method e.g. jwk, dht, web
   String name;
 
   /// The function to resolve a DID URI using this method.
-  DidResolutionResult Function(String) resolve;
+  Future<DidResolutionResult> Function(String) resolve;
 
   /// Constructs a [DidMethodResolver] with a given [name] and [resolve] function.
   DidMethodResolver({required this.name, required this.resolve});
+
+  Future<DidDereferenceResult> dereference(String didUri) async {
+    final didResolutionResult = await resolve(didUri);
+
+    if (didResolutionResult.hasError()) {
+      return DidDereferenceResult.withError(
+        didResolutionResult.didResolutionMetadata.error!,
+      );
+    }
+
+    if (!didUri.contains('#')) {
+      return DidDereferenceResult(
+        contentStream: didResolutionResult.didDocument as DidResource,
+        contentMetadata: didResolutionResult.didDocumentMetadata,
+      );
+    }
+
+    final resource = didResolutionResult.didDocument!.getResourceById(didUri);
+    if (resource != null) {
+      return DidDereferenceResult(contentStream: resource);
+    } else {
+      return DidDereferenceResult.withError('notFound');
+    }
```

Comment:
nit. 
```dart
return resource != null ? 
  DidDereferenceResult(contentStream: resource) :
  DidDereferenceResult.withError('notFound');
```

Reply:
love it. changing now

Reply:
done: https://github.com/TBD54566975/web5-dart/commit/fa5dbae2693ba6d8664b74fffff21fbb83807b1e